File: server/src/main/java/io/fnx/backend/domain/UserEntity.java
package io.fnx.backend.domain;

import com.googlecode.objectify.Key;
import com.googlecode.objectify.annotation.*;
import io.fnx.backend.tools.auth.Principal;
import io.fnx.backend.tools.auth.PrincipalRole;
import io.fnx.backend.tools.authorization.OwnedEntity;
import org.joda.time.DateTime;

import java.util.Collections;
import java.util.List;

import static io.fnx.backend.tools.ofy.OfyUtils.idToKey;

@Cache
@Entity
@Unindex
public class UserEntity implements Principal, OwnedEntity<UserEntity> {

    @Id
    private Long id;

    @Index
    private String email;

    private String firstName;

	@Index
	private String lastName;

    private List<Role> roles;

    @JsonIgnore
    private String passwordHash;
	
	@JsonIgnore
    private DateTime passwordTokenValidTill;

	@JsonIgnore
    private String passwordToken;
	
	@Index
	private String avatarUrl;
	
    public static Key<UserEntity> createKey(Long id) {
        return idToKey(UserEntity.class, id);
    }

    public Key<UserEntity> getKey() {
        return createKey(id);
    }

    @Override
    public Key<? extends Principal> getPrincipalKey() {
        return getKey();
    }

    @Override
    public List<? extends PrincipalRole> getUserRoles() {
        return roles;
    }

    @Override
    public Key<UserEntity> getOwnerKey() {
        return getKey(); //users "own" themselves
    }

    @Override
    public String toString() {
        return "UserEntity{" +
                "id=" + id +
                ", email='" + email + '\'' +
                ", roles=" + roles +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof UserEntity)) return false;

        UserEntity that = (UserEntity) o;

        return id != null ? id.equals(that.id) : that.id == null;
    }

    @Override
    public int hashCode() {
        return id != null ? id.hashCode() : 0;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public void setPasswordHash(String passwordHash) {
        this.passwordHash = passwordHash;
    }

    public String getPasswordHash() {
        return passwordHash;
    }

    public String getFullName() {
    	if (lastName == null) {
    		if (firstName == null) return "";
    		return firstName;
	    }
	    if (firstName == null) {
    		return lastName;
	    }
	    return firstName+" "+lastName;
    }

	public String getFirstName() {
		return firstName;
	}

	public void setFirstName(String firstName) {
		this.firstName = firstName;
	}

	public String getLastName() {
		return lastName;
	}

	public void setLastName(String lastName) {
		this.lastName = lastName;
	}

	public DateTime getPasswordTokenValidTill() {
		return passwordTokenValidTill;
	}

	public void setPasswordTokenValidTill(DateTime passwordTokenValidTill) {
		this.passwordTokenValidTill = passwordTokenValidTill;
	}

	public String getPasswordToken() {
		return passwordToken;
	}

	public void setPasswordToken(String passwordToken) {
		this.passwordToken = passwordToken;
	}
	
	public String getAvatarUrl() {
		return avatarUrl;
	}

	public void setAvatarUrl(String avatarUrl) {
		this.avatarUrl = avatarUrl;
	}

    public List<Role> getRoles() {
        return roles;
    }

    public void setRoles(List<Role> roles) {
        this.roles = roles;
    }

    public void setSingleRole(Role role) {
        this.roles = Collections.singletonList(role);
    }
}

File: api_client/lib/model/user_entity.dart
part of api_client.api;

class UserEntity {
  
  int id = null;
  

  String email = null;
  

  String firstName = null;
  

  String lastName = null;
  

  String password = null;
  

  List<Role> roles = [];
  

  String avatarUrl = null;
  
  UserEntity();

  @override
  String toString() {
    return 'UserEntity[id=$id, email=$email, firstName=$firstName, lastName=$lastName, password=$password, roles=$roles, avatarUrl=$avatarUrl, ]';
  }

  UserEntity.fromJson(Map<String, dynamic> json) {
    if (json == null) return;
    id =
        json['id']
    ;
    email =
        json['email']
    ;
    firstName =
        json['firstName']
    ;
    lastName =
        json['lastName']
    ;
    password =
        json['password']
    ;
    roles =
      Role.listFromJson(json['roles'])
;
    avatarUrl =
        json['avatarUrl']
    ;
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'email': email,
      'firstName': firstName,
      'lastName': lastName,
      'password': password,
      'roles': roles,
      'avatarUrl': avatarUrl
     };
  }

  static List<UserEntity> listFromJson(List<dynamic> json) {
    return json == null ? new List<UserEntity>() : json.map((value) => new UserEntity.fromJson(value)).toList();
  }

  static Map<String, UserEntity> mapFromJson(Map<String, Map<String, dynamic>> json) {
    var map = new Map<String, UserEntity>();
    if (json != null && json.length > 0) {
      json.forEach((String key, Map<String, dynamic> value) => map[key] = new UserEntity.fromJson(value));
    }
    return map;
  }
}

